network/router_test.go

```diff
@@ -63,6 +63,61 @@ func testRouter(t *testing.T, fac routerFactory) {
 	}
 }
 
+// Test connection of multiple Hosts and sending messages back and forth
+// also tests for the counterIO interface that it works well
+func TestRouterErrorHandling(t *testing.T) {
+	h1, err1 := NewTestRouterTCP(2109)
+	h2, err2 := NewTestRouterTCP(2110)
+	if err1 != nil || err2 != nil {
+		t.Fatal("Could not setup hosts")
+	}
+
+	go h1.Start()
+	go h2.Start()
+
+	defer func() {
+		h1.Stop()
+	}()
+
+	// tests the setting error handler
+	require.NotNil(t, h1.errorHandlers)
+	if len(h1.errorHandlers) != 0 {
+		t.Error("errorHandlers should start empty")
+	}
+	errHandlerCalled := make(chan bool, 1)
+	errHandler := func(remote *ServerIdentity) {
+		errHandlerCalled <- true
+	}
+	h1.AddErrorHandler(errHandler)
+	if len(h1.errorHandlers) != 1 {
+		t.Error("errorHandlers should now hold one function")
+	}
+
+	//register handlers
+	proc := &simpleMessageProc{t, make(chan SimpleMessage)}
+	h1.RegisterProcessor(proc, SimpleMessageType)
+	h2.RegisterProcessor(proc, SimpleMessageType)
+
+	msgSimple := &SimpleMessage{3}
+	err := h1.Send(h2.ServerIdentity, msgSimple)
+	require.Nil(t, err)
+	decoded := <-proc.relay
+	assert.Equal(t, 3, decoded.I)
+	assert.Nil(t, h2.Send(h1.ServerIdentity, msgSimple))
+	decoded = <-proc.relay
+
+	//stop node 2
+	h2.Stop()
+	time.Sleep(250 * time.Millisecond)
```

Comment:
why do you sleep here ?
If it's because it takes time to detect the error and propagate the error to the handler, you should be using channels like below:
```go
select {
    case <-errHandlerCalled:
       // all good
    case <-time.After(100 * time.Millisecond):
      t.Error("Error handler should ... ")
}
```